File: utils/formatters.py
"""
formatters.py - Display Formatting Utilities
==============================================

Pure functions for formatting bandwidth rates and data totals.

CRITICAL DISTINCTION:
    format_bandwidth_rate() → For dashboard bandwidth card, chart Y-axis (RATE in Mbps)
    format_total_data()     → For device "Total Sent/Received" columns (CUMULATIVE in MB/GB)
"""


def format_bandwidth_rate(bytes_per_second: float) -> str:
    """
    Format bandwidth RATE for display.

    Input:  bytes per second (from packet capture / BandwidthCalculator)
    Output: Human-readable string with Mbps / Kbps / B/s

    CRITICAL: This is for RATE (bps), not total data transferred.

    Examples:
        375_000  → "3.0 Mbps"   (YouTube HD)
        12_500   → "100.0 Kbps" (light browsing)
        50       → "50 B/s"     (idle)
        0        → "0 B/s"
    """
    if bytes_per_second <= 0:
        return "0 B/s"

    # Convert bytes/sec → megabits/sec
    mbps = (bytes_per_second * 8) / 1_000_000

    if mbps >= 1.0:
        # >= 1 Mbps: Show as "X.X Mbps"
        return f"{mbps:.1f} Mbps"

    # Convert to kilobits/sec
    kbps = (bytes_per_second * 8) / 1_000

    if kbps >= 1.0:
        # >= 1 Kbps: Show as "XX.X Kbps"
        return f"{kbps:.1f} Kbps"

    # Show raw bytes/sec
    return f"{bytes_per_second:.0f} B/s"


def format_total_data(total_bytes: int) -> str:
    """
    Format total data transferred (cumulative).

    Input:  total bytes (sum of all packets)
    Output: Human-readable string with GB / MB / KB / B

    CRITICAL: This is for TOTAL data, not rate.

    Examples:
        1_500_000_000 → "1.50 GB"
        150_500_000   → "150.5 MB"
        45_200        → "45.2 KB"
        500           → "500 B"
    """
    if total_bytes is None or total_bytes < 0:
        total_bytes = 0

    if total_bytes >= 1_000_000_000:
        gb = total_bytes / 1_000_000_000
        return f"{gb:.2f} GB"
    elif total_bytes >= 1_000_000:
        mb = total_bytes / 1_000_000
        return f"{mb:.1f} MB"
    elif total_bytes >= 1_000:
        kb = total_bytes / 1_000
        return f"{kb:.1f} KB"
    else:
        return f"{total_bytes} B"


def bytes_to_mbps(bytes_per_second: float) -> float:
    """
    Convert bytes per second to megabits per second.

    Args:
        bytes_per_second: Raw bytes/second value

    Returns:
        Mbps as a float, rounded to 4 decimal places
    """
    if bytes_per_second <= 0:
        return 0.0
    return round((bytes_per_second * 8) / 1_000_000, 4)

File: frontend/js/utils/formatters.js
/**
 * formatters.js - Display Formatting Utilities
 * ==============================================
 * Pure functions for formatting bytes, dates, and other values.
 */

export function formatBytes(bytes, decimals = 2) {
  if (bytes === 0) return '0 B';
  if (!bytes || isNaN(bytes)) return '-- B';
  const k = 1024;
  const sizes = ['B', 'KB', 'MB', 'GB', 'TB'];
  const i = Math.floor(Math.log(Math.abs(bytes)) / Math.log(k));
  return parseFloat((bytes / Math.pow(k, i)).toFixed(decimals)) + ' ' + sizes[Math.min(i, sizes.length - 1)];
}

/**
 * Format a bandwidth RATE in bytes/sec to human-readable Mbps/Kbps/B/s.
 * This is the network-standard display: bytes → bits → Mbps.
 * Use this for dashboard bandwidth cards and device current rates.
 */
export function formatBandwidthRate(bytesPerSecond) {
  if (!bytesPerSecond || bytesPerSecond <= 0) return '0 B/s';
  const mbps = (bytesPerSecond * 8) / 1_000_000;
  if (mbps >= 1) return mbps.toFixed(1) + ' Mbps';
  const kbps = (bytesPerSecond * 8) / 1_000;
  if (kbps >= 1) return kbps.toFixed(1) + ' Kbps';
  return Math.round(bytesPerSecond) + ' B/s';
}

/**
 * Format a value that is already in Mbps into a human-readable string.
 * Always uses Mbps to stay consistent with the dashboard card unit.
 * Sub-Mbps values are shown as fractional Mbps (e.g. "0.30 Mbps" not "300 Kbps").
 */
export function formatMbps(mbps) {
  if (mbps == null || isNaN(mbps) || mbps <= 0) return '0 B/s';
  if (mbps >= 1)     return mbps.toFixed(1) + ' Mbps';
  if (mbps >= 0.01)  return mbps.toFixed(2) + ' Mbps';   // 0.01–0.99 Mbps
  if (mbps >= 0.001) return mbps.toFixed(3) + ' Mbps';   // 0.001–0.009 Mbps
  // Extremely small — fall back to Kbps / B/s for readability
  const kbps = mbps * 1000;
  if (kbps >= 1)   return kbps.toFixed(1) + ' Kbps';
  const bytesPerSec = (mbps * 1_000_000) / 8;
  if (bytesPerSec >= 1) return Math.round(bytesPerSec) + ' B/s';
  return '0 B/s';
}

/**
 * Split a bandwidth rate (bytes/sec) into {value, unit} for card display.
 * Matches formatBandwidthRate() thresholds exactly.
 */
export function splitBandwidthRate(bytesPerSecond) {
  if (!bytesPerSecond || bytesPerSecond <= 0) return { value: '0', unit: 'B/s' };
  const mbps = (bytesPerSecond * 8) / 1_000_000;
  if (mbps >= 1) return { value: mbps.toFixed(1), unit: 'Mbps' };
  const kbps = (bytesPerSecond * 8) / 1_000;
  if (kbps >= 1) return { value: kbps.toFixed(1), unit: 'Kbps' };
  return { value: Math.round(bytesPerSecond).toString(), unit: 'B/s' };
}

export function formatTimestamp(ts) {
  if (!ts) return '--:--';
  // Normalise "YYYY-MM-DD HH:MM:SS" → ISO 8601 "YYYY-MM-DDTHH:MM:SS"
  // so browsers that only accept the 'T' separator don't return NaN.
  const d = new Date(typeof ts === 'string' ? ts.replace(' ', 'T') : ts);
  if (isNaN(d.getTime())) return '--:--';
  // Show seconds for sub-minute data (e.g. 10s/30s buckets) so that chart
  // x-axis labels are distinct within the same minute.
  if (d.getSeconds() !== 0) {
    return d.toLocaleTimeString('en-US', { hour: '2-digit', minute: '2-digit', second: '2-digit', hour12: false });
  }
  return d.toLocaleTimeString('en-US', { hour: '2-digit', minute: '2-digit', hour12: false });
}

export function formatRelativeTime(ts) {
  if (!ts) return 'unknown';
  const seconds = Math.floor((Date.now() - new Date(typeof ts === 'string' ? ts.replace(' ', 'T') : ts).getTime()) / 1000);
  if (seconds < 0 || seconds < 60) return 'just now';
  if (seconds < 3600) return `${Math.floor(seconds / 60)}m ago`;
  if (seconds < 86400) return `${Math.floor(seconds / 3600)}h ago`;
  return `${Math.floor(seconds / 86400)}d ago`;
}

export function escapeHtml(text) {
  if (!text) return '';
  const div = document.createElement('div');
  div.textContent = text;
  return div.innerHTML;
}

export function severityClass(severity) {
  const map = { critical: 'critical', high: 'critical', warning: 'warning', medium: 'warning', info: 'info', low: 'info' };
  return map[severity?.toLowerCase()] || 'info';
}
